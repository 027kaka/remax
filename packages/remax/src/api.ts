export const getApp = (...args: any): any => {};
export const addCardAuth = (...args: any): any => {};
export const addPhoneContact = (...args: any): any => {};
export const alert = (...args: any): any => {};
export const canIUse = (...args: any): any => {};
export const chooseAlipayContact = (...args: any): any => {};
export const chooseCity = (...args: any): any => {};
export const chooseContact = (...args: any): any => {};
export const chooseImage = (...args: any): any => {};
export const chooseLocation = (...args: any): any => {};
export const choosePhoneContact = (...args: any): any => {};
export const clearStorage = (...args: any): any => {};
export const clearStorageSync = (...args: any): any => {};
export const closeBluetoothAdapter = (...args: any): any => {};
export const closeSocket = (...args: any): any => {};
export const compressImage = (...args: any): any => {};
export const confirm = (...args: any): any => {};
export const connectBLEDevice = (...args: any): any => {};
export const connectSocket = (...args: any): any => {};
export const createAnimation = (...args: any): any => {};
export const createCanvasContext = (...args: any): any => {};
export const createIntersectionObserver = (...args: any): any => {};
export const createMapContext = (...args: any): any => {};
export const createSelectorQuery = (...args: any): any => {};
export const createWebViewContext = (...args: any): any => {};
export const datePicker = (...args: any): any => {};
export const disconnectBLEDevice = (...args: any): any => {};
export const downloadFile = (...args: any): any => {};
export const getAuthCode = (...args: any): any => {};
export const getAuthUserInfo = (...args: any): any => {};
export const getBatteryInfo = (...args: any): any => {};
export const getBatteryInfoSync = (...args: any): any => {};
export const getBeacons = (...args: any): any => {};
export const getBLEDeviceCharacteristics = (...args: any): any => {};
export const getBLEDeviceServices = (...args: any): any => {};
export const getBluetoothAdapterState = (...args: any): any => {};
export const getBluetoothDevices = (...args: any): any => {};
export const getClipboard = (...args: any): any => {};
export const getConnectedBluetoothDevices = (...args: any): any => {};
export const getFileInfo = (...args: any): any => {};
export const getImageInfo = (...args: any): any => {};
export const getLocation = (...args: any): any => {};
export const getNetworkType = (...args: any): any => {};
export const getPhoneNumber = (...args: any): any => {};
export const getRunData = (...args: any): any => {};
export const getRunScene = (...args: any): any => {};
export const getSavedFileInfo = (...args: any): any => {};
export const getSavedFileList = (...args: any): any => {};
export const getScreenBrightness = (...args: any): any => {};
export const getServerTime = (...args: any): any => {};
export const getSetting = (...args: any): any => {};
export const getStorage = (...args: any): any => {};
export const getStorageInfo = (...args: any): any => {};
export const getStorageInfoSync = (...args: any): any => {};
export const getStorageSync = (...args: any): any => {};
export const getSystemInfo = (...args: any): any => {};
export const getSystemInfoSync = (...args: any): any => {};
export const getTitleColor = (...args: any): any => {};
export const getUpdateManager = (...args: any): any => {};
export const hideAddToDesktopMenu = (...args: any): any => {};
export const hideAllAddToDesktopMenu = (...args: any): any => {};
export const hideAllFavoriteMenu = (...args: any): any => {};
export const hideBackHome = (...args: any): any => {};
export const hideFavoriteMenu = (...args: any): any => {};
export const hideKeyboard = (...args: any): any => {};
export const hideLoading = (...args: any): any => {};
export const hideNavigationBarLoading = (...args: any): any => {};
export const hideShareMenu = (...args: any): any => {};
export const hideTabBar = (...args: any): any => {};
export const hideTabBarRedDot = (...args: any): any => {};
export const hideToast = (...args: any): any => {};
export const loadFontFace = (...args: any): any => {};
export const makePhoneCall = (...args: any): any => {};
export const multiLevelSelect = (...args: any): any => {};
export const navigateBack = (...args: any): any => {};
export const navigateBackMiniProgram = (...args: any): any => {};
export const navigateTo = (...args: any): any => {};
export const navigateToMiniProgram = (...args: any): any => {};
export const notifyBLECharacteristicValueChange = (...args: any): any => {};
export const offAccelerometerChange = (...args: any): any => {};
export const offBLECharacteristicValueChange = (...args: any): any => {};
export const offBLEConnectionStateChanged = (...args: any): any => {};
export const offBluetoothAdapterStateChange = (...args: any): any => {};
export const offBluetoothDeviceFound = (...args: any): any => {};
export const offCompassChange = (...args: any): any => {};
export const offGyroscopeChange = (...args: any): any => {};
export const offMemoryWarning = (...args: any): any => {};
export const offNetworkStatusChange = (...args: any): any => {};
export const offSocketClose = (...args: any): any => {};
export const offSocketError = (...args: any): any => {};
export const offSocketMessage = (...args: any): any => {};
export const offSocketOpen = (...args: any): any => {};
export const offUserCaptureScreen = (...args: any): any => {};
export const onAccelerometerChange = (...args: any): any => {};
export const onBeaconServiceChange = (...args: any): any => {};
export const onBeaconUpdate = (...args: any): any => {};
export const onBLECharacteristicValueChange = (...args: any): any => {};
export const onBLEConnectionStateChanged = (...args: any): any => {};
export const onBluetoothAdapterStateChange = (...args: any): any => {};
export const onBluetoothDeviceFound = (...args: any): any => {};
export const onCompassChange = (...args: any): any => {};
export const onGyroscopeChange = (...args: any): any => {};
export const onMemoryWarning = (...args: any): any => {};
export const onNetworkStatusChange = (...args: any): any => {};
export const onSocketClose = (...args: any): any => {};
export const onSocketError = (...args: any): any => {};
export const onSocketMessage = (...args: any): any => {};
export const onSocketOpen = (...args: any): any => {};
export const onUserCaptureScreen = (...args: any): any => {};
export const openBluetoothAdapter = (...args: any): any => {};
export const openCardDetail = (...args: any): any => {};
export const openCardList = (...args: any): any => {};
export const openKBVoucherDetail = (...args: any): any => {};
export const openLocation = (...args: any): any => {};
export const openMerchantCardList = (...args: any): any => {};
export const openMerchantTicketList = (...args: any): any => {};
export const openMerchantVoucherList = (...args: any): any => {};
export const openSetting = (...args: any): any => {};
export const openTicketDetail = (...args: any): any => {};
export const openTicketList = (...args: any): any => {};
export const openVoucherDetail = (...args: any): any => {};
export const openVoucherList = (...args: any): any => {};
export const optionsSelect = (...args: any): any => {};
export const pageScrollTo = (...args: any): any => {};
export const previewImage = (...args: any): any => {};
export const prompt = (...args: any): any => {};
export const readBLECharacteristicValue = (...args: any): any => {};
export const redirectTo = (...args: any): any => {};
export const reLaunch = (...args: any): any => {};
export const removeSavedFile = (...args: any): any => {};
export const removeStorage = (...args: any): any => {};
export const removeStorageSync = (...args: any): any => {};
export const removeTabBarBadge = (...args: any): any => {};
export const reportAnalytics = (...args: any): any => {};
export const request = (...args: any): any => {};
export const rsa = (...args: any): any => {};
export const saveFile = (...args: any): any => {};
export const saveImage = (...args: any): any => {};
export const scan = (...args: any): any => {};
export const SDKVersion = (...args: any): any => {};
export const sendSocketMessage = (...args: any): any => {};
export const setBackgroundColor = (...args: any): any => {};
export const setBackgroundTextStyle = (...args: any): any => {};
export const setCanPullDown = (...args: any): any => {};
export const setClipboard = (...args: any): any => {};
export const setKeepScreenOn = (...args: any): any => {};
export const setNavigationBar = (...args: any): any => {};
export const setOptionMenu = (...args: any): any => {};
export const setScreenBrightness = (...args: any): any => {};
export const setStorage = (...args: any): any => {};
export const setStorageSync = (...args: any): any => {};
export const setTabBarBadge = (...args: any): any => {};
export const setTabBarItem = (...args: any): any => {};
export const setTabBarStyle = (...args: any): any => {};
export const showActionSheet = (...args: any): any => {};
export const showAuthGuide = (...args: any): any => {};
export const showLoading = (...args: any): any => {};
export const showNavigationBarLoading = (...args: any): any => {};
export const showSharePanel = (...args: any): any => {};
export const showTabBar = (...args: any): any => {};
export const showTabBarRedDot = (...args: any): any => {};
export const showToast = (...args: any): any => {};
export const startBeaconDiscovery = (...args: any): any => {};
export const startBluetoothDevicesDiscovery = (...args: any): any => {};
export const startPullDownRefresh = (...args: any): any => {};
export const startZMVerify = (...args: any): any => {};
export const stopBeaconDiscovery = (...args: any): any => {};
export const stopBluetoothDevicesDiscovery = (...args: any): any => {};
export const stopPullDownRefresh = (...args: any): any => {};
export const switchTab = (...args: any): any => {};
export const textRiskIdentification = (...args: any): any => {};
export const tradePay = (...args: any): any => {};
export const uploadFile = (...args: any): any => {};
export const vibrate = (...args: any): any => {};
export const vibrateLong = (...args: any): any => {};
export const vibrateShort = (...args: any): any => {};
export const watchShake = (...args: any): any => {};
export const writeBLECharacteristicValue = (...args: any): any => {};
