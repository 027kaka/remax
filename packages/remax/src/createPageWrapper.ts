import * as React from 'react';

export default function createPageWrapper(Page: React.ComponentType, query: object) {
  return class PageWrapper extends React.Component {
    instance: any = null;

    componentDidShow() {
      if (this.instance && this.instance.componentDidShow) {
        this.instance.componentDidShow();
      }
    }

    render() {
      const props: any = {
        location: {
          query: query || {},
        },
      };
      // function component doesn't have ref
      if (typeof Page !== 'function') {
        props.ref = (node: any) => (this.instance = node);
      }
      return React.createElement(Page, props);
    }
  };
}
