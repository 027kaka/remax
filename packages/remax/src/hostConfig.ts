import * as scheduler from '@remax/scheduler';

const {
  unstable_scheduleCallback: scheduleDeferredCallback,
  unstable_cancelCallback: cancelDeferredCallback,
  unstable_shouldYield: shouldYield,
  unstable_now: now,
} = scheduler;

const REMAX_ROOT = '$$REMAX_ROOT';
const REMAX_ROOT_BACKUP = '$$REMAX_ROOT_BACKUP';
const REMAX_METHOD = '$$REMAX_METHOD';
const TYPE_TEXT = Symbol('text');

let instanceCount = 0;

// 缓冲一下要 set 的 Data
// 如果有值说明还没 set，直接改变他的值
// 真正 setData 后把 lastData 置空
let lastData: any = null;

function setData(rootContext: any) {
  function clone(item: any) {
    let result: any = {};
    if (Array.isArray(item)) {
      result = item.map(item => clone(item));
    } else if (typeof item === 'object') {
      for (const key of Object.keys(item)) {
        if (key !== 'rootContext') {
          result[key] = clone(item[key]);
        }
      }
    } else {
      result = item;
    }

    return result;
  }

  const pureObject = clone(rootContext[REMAX_ROOT_BACKUP]);

  if (lastData) {
    // 更新了 lastData 等待 setData 发生即可
    lastData = pureObject;
  } else {
    lastData = pureObject;
    const startTime = new Date().getTime();
    rootContext.setData(
      {
        [REMAX_ROOT]: lastData,
      },
      () => {
        if (process.env.NODE_ENV !== 'production') {
          console.log(`setData => 回调时间：${new Date().getTime() - startTime}ms`);
        }
      },
    );

    lastData = null;
  }
}

function processProps(newProps: any, rootContext: any, id: number) {
  const props: any = {};
  for (const propKey of Object.keys(newProps)) {
    if (typeof newProps[propKey] === 'function') {
      const contextKey = `${REMAX_METHOD}_${id}_${propKey}`;
      rootContext[contextKey] = newProps[propKey];
      props[propKey] = contextKey;
    } else if (propKey === 'children') {
      // pass
    } else {
      props[propKey] = newProps[propKey];
    }
  }
  return props;
}

const rootHostContext = {};
const childHostContext = {};

export default {
  now,

  getRootHostContext: () => {
    return rootHostContext;
  },

  shouldSetTextContent(type: any, props: any) {
    return false;
  },

  prepareForCommit: () => {},

  resetAfterCommit: () => {},

  getChildHostContext: () => {
    return childHostContext;
  },

  prepareUpdate() {
    return true;
  },

  commitTextUpdate(textInstance: any, oldText: string, newText: string) {
    textInstance.text = newText;
    setData(textInstance.rootContext);
  },

  createInstance: (type: string, newProps: any, rootContainerInstance: any, _currentHostContext: any) => {
    const rootContext = rootContainerInstance;
    const id = instanceCount;
    instanceCount += 1;

    const props = processProps(newProps, rootContext, id);

    const ins = {
      type: type === 'div' ? 'view' : type,
      props,
      children: [],
      rootContext,
      id,
    };

    return ins;
  },

  createTextInstance(text: string) {
    return {
      type: TYPE_TEXT,
      text,
    };
  },

  commitUpdate(targetIns: any, updatePayload: any, type: string, oldProps: any, newProps: any) {
    const props = processProps(newProps, targetIns.rootContext, targetIns.id);

    targetIns.props = props;

    setData(targetIns.rootContext);
  },

  appendInitialChild: (parent: any, child: any) => {
    child.rootContext = parent.rootContext;
    parent.children.push(child);
  },

  appendChild(parent: any, child: any) {
    child.rootContext = parent.rootContext;
    parent.children.push(child);
  },

  insertBefore(parent: any, child: any) {
    child.rootContext = parent.rootContext;
    parent.children.unshift(child);
  },

  finalizeInitialChildren: () => {
    return false;
  },

  supportsMutation: true,

  appendChildToContainer: (_parent: any, child: any) => {
    let parent: any = null;
    if (_parent._rootContainer) {
      // append to root
      parent = {
        type: 'root',
        children: [],
        rootContext: _parent,
      };
    }

    parent.children.push(child);

    child.rootContext[REMAX_ROOT_BACKUP] = child.rootContext[REMAX_ROOT_BACKUP] || [];
    child.rootContext[REMAX_ROOT_BACKUP].push(parent);
    setData(child.rootContext);
  },

  removeChild(parentInstance: any, child: any) {
    parentInstance.children.splice(parentInstance.indexOf(child), 1);
  },

  schedulePassiveEffects: scheduleDeferredCallback,
  cancelPassiveEffects: cancelDeferredCallback,
  shouldYield,
  scheduleDeferredCallback,
  cancelDeferredCallback,
};
