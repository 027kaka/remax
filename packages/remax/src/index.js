import ReactReconciler from 'react-reconciler';
import React from 'react';
import {
  useState,
} from 'react';

const REACT_MINI_APP_ROOT = '$$REACT_MINI_APP_ROOT';
const REACT_MINI_APP_ROOT_BACKUP = '$$REACT_MINI_APP_ROOT_BACKUP';
const REACT_MINI_APP_METHOD = '$$REACT_MINI_APP_METHOD';
const TYPE_TEXT = Symbol('text');

let instanceCount = 0;

function setData(rootContext) {
  function clone(item) {
    let result = {};
    if (Array.isArray(item)) {
      result = item.map(item => clone(item));
    } else if (typeof item === 'object') {
      for (const key of Object.keys(item)) {
        if (key !== 'rootContext') {
          result[key] = clone(item[key]);
        }
      }
    } else {
      result = item;
    }

    return result;
  }

  const pureObject = clone(rootContext[REACT_MINI_APP_ROOT_BACKUP]);

  rootContext.setData({
    [REACT_MINI_APP_ROOT]: [pureObject],
  });
}

function processProps(newProps, rootContext, id) {
  const props = {};
  for (const propKey of Object.keys(newProps)) {
    if (typeof newProps[propKey] === 'function') {
      const contextKey = `${REACT_MINI_APP_METHOD}_${id}_${propKey}`;
      rootContext[contextKey] = newProps[propKey];
      props[propKey] = contextKey;
    } else if (propKey === 'children') {
      // pass
    } else {
      props[propKey] = newProps[propKey];
    }
  }
  return props;
}

const rootHostContext = {};
const childHostContext = {};

const hostConfig = {
  now: Date.now,

  getRootHostContext: () => {
    return rootHostContext;
  },
  shouldSetTextContent: (type, props) => {
    return typeof props.children === 'string' || typeof props.children === 'number';
  },
  prepareForCommit: () => {},
  resetAfterCommit: () => {},
  getChildHostContext: () => {
    return childHostContext;
  },
  prepareUpdate(domElement, oldProps, newProps) {
    return true;
  },
  commitTextUpdate(textInstance, oldText, newText) {
    textInstance.text = newText;
    setData(textInstance.rootContext);
  },

  createInstance: (type, newProps, rootContainerInstance, _currentHostContext, workInProgress) => {
    const rootContext = rootContainerInstance;
    const id = instanceCount;
    instanceCount += 1;

    const props = processProps(newProps, rootContext, id);

    const ins = {
      type: type === 'div' ? 'view' : type,
      props,
      children: [],
      rootContext,
      id,
    };

    return ins;
  },
  createTextInstance: (text) => {
    return {
      type: TYPE_TEXT,
      text,
    };
  },
  commitUpdate(targetIns, updatePayload, type, oldProps, newProps) {
    const props = processProps(newProps, targetIns.rootContext, targetIns.id);

    targetIns.props = props;

    setData(targetIns.rootContext);
  },
  appendInitialChild: (parent, child) => {
    child.rootContext = parent.rootContext;
    parent.children.push(child);
  },
  appendChild(parent, child) {
    child.rootContext = parent.rootContext;
    parent.children.push(child);
  },
  finalizeInitialChildren: () => {},
  supportsMutation: true,
  appendChildToContainer: (_parent, child) => {
    let parent = null;
    if (_parent._rootContainer) {
      // append to root
      parent = {
        type: 'view',
        children: [],
        rootContext: _parent,
      };
    }


    parent.children.push(child);

    child.rootContext[REACT_MINI_APP_ROOT_BACKUP] = parent;
    setData(child.rootContext);
  },

  removeChild(parentInstance, child) {
    parentInstance.children.splice(parentInstance.indexOf(child), 1);
  },
};

const ReactReconcilerInst = ReactReconciler(hostConfig);
export default {
  ...React,
  React,
  render: (reactElement, miniAppContext, callback) => {
    // Create a root Container if it doesnt exist
    if (!miniAppContext._rootContainer) {
      miniAppContext._rootContainer = ReactReconcilerInst.createContainer(miniAppContext, false);
    }

    // update the root Container
    return ReactReconcilerInst.updateContainer(reactElement, miniAppContext._rootContainer, null, callback);
  },
};
