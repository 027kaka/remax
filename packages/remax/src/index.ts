export { default as render } from './render';
export { default as createAppConfig } from './createAppConfig';
export { default as createPageConfig } from './createPageConfig';
export { PageProps, LifecycleHooks } from './createPageWrapper';
export { default as Platform } from './Platform';
