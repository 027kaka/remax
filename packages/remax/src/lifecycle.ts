import capitalize from './utils/capitalize';
import lowercase from './utils/lowercase';

export type Callback = (...args: any[]) => any;

export enum Lifecycle {
  load = 'load',
  show = 'show',
  hide = 'hide',
  ready = 'ready',
  pullDownRefresh = 'pullDownRefresh',
  reachBottom = 'reachBottom',
  pageScroll = 'pageScroll',
  shareAppMessage = 'shareAppMessage',
  titleClick = 'titleClick',
  optionMenuClick = 'optionMenuClick',
  popMenuClick = 'popMenuClick',
  pullIntercept = 'pullIntercept',
  back = 'back',
  keyboardHeight = 'keyboardHeight',
  tabItemTap = 'tabItemTap',
  beforeTabItemTap = 'beforeTabItemTap',
  resize = 'resize',
  unload = 'unload',
}

export enum AppLifecycle {
  launch = 'launch',
  show = 'show',
  hide = 'hide',
  error = 'error',
  shareAppMessage = 'shareAppMessage',
  pageNotFound = 'pageNotFound',
}

export function lifeCycleName(name: string): Lifecycle {
  if (name.startsWith('before')) {
    return name as Lifecycle;
  }
  return lowercase(name.slice(2)) as Lifecycle;
}

export function callbackName(name: string) {
  if (name.startsWith('before')) {
    return capitalize(name);
  }
  return 'on' + capitalize(name);
}

export function registerLifecycle(instance: any, method: Lifecycle | AppLifecycle, callback: Callback) {
  return instance.registerLifecycle(method, callback);
}

const ENTRY_INFO = __REMAX_ENTRY_INFO__;
const PAGE_EVENTS = __REMAX_PAGE_EVENTS__;
const APP_EVENTS = __REMAX_APP_EVENTS__;

export function pageEvents(entry: string) {
  const entryInfo = ENTRY_INFO.find((e: any) => e.name === entry);
  return ((entryInfo?.modules || []) as string[]).reduce<string[]>((acc, cur) => {
    return [...acc, ...(PAGE_EVENTS[cur] || [])];
  }, []);
}

export function appEvents(): string[] {
  return APP_EVENTS;
}
