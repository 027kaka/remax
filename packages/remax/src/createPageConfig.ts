import * as React from 'react';
import createPageWrapper from './createPageWrapper';
import render from './render';

export default function createPageConfig(Page: React.ComponentType<any>) {
  return {
    data: {
      $$REMAX_ROOT: [],
    },

    wrapper: null as any,

    onLoad(query: any) {
      const PageWrapper = createPageWrapper(Page, query);

      this.wrapper = render(React.createElement(PageWrapper), this);
    },

    onUnload() {
      render(null, this);
      this.wrapper = null;
    },

    onShow() {
      this.wrapper.onShow();
    },

    onHide() {
      this.wrapper.onHide();
    },

    onPullDownRefresh() {
      this.wrapper.onPullDownRefresh();
    },

    onReachBottom() {
      this.wrapper.onReachBottom();
    },

    onPageScroll() {
      this.wrapper.onPageScroll();
    },

    onShareAppMessage() {
      this.wrapper.onShareAppMessage();
    },

    onTitleClick() {
      this.wrapper.onTitleClick();
    },

    onOptionMenuClick() {
      this.wrapper.onOptionMenuClick();
    },

    onPopMenuClick() {
      this.wrapper.onPopMenuClick();
    },

    onPullIntercept() {
      this.wrapper.onPullIntercept();
    },
  };
}
