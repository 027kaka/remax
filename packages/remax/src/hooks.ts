import { useLayoutEffect, useContext, DependencyList } from 'react';
import { registerLifecycle, Callback, lifeCycleName } from './lifecycle';
import PageInstanceContext from './PageInstanceContext';
import AppInstanceContext from './AppInstanceContext';
import nativeEffect, { Listener } from './nativeEffect';

/**
 *  Page Hooks
 */

export function usePageInstance() {
  return useContext(PageInstanceContext);
}

// eslint-disable-next-line @typescript-eslint/camelcase
export function unstable_useNativeEffect(listener: Listener, deps?: DependencyList) {
  useLayoutEffect(() => {
    return nativeEffect.connect(listener, !!deps);
  }, deps);
}

export function useQuery<Q extends {} = { [name: string]: string }>(): Q {
  const pageInstance: any = useContext(PageInstanceContext);
  return pageInstance.query;
}

export function usePageEvent(eventName: string, callback: Callback) {
  const pageInstance = useContext(PageInstanceContext);
  const lifeCycle = lifeCycleName(eventName);
  useLayoutEffect(() => {
    return registerLifecycle(pageInstance, lifeCycle, callback);
  });
}

/**
 * App Hooks
 */

export function useAppEvent(eventName: string, callback: Callback) {
  const lifeCycle = lifeCycleName(eventName);
  useLayoutEffect(() => {
    return registerLifecycle(AppInstanceContext, lifeCycle, callback);
  });
}
