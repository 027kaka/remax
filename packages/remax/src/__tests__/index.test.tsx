import * as React from 'react';
import { render, View } from '../../src';
import { REMAX_ROOT_BACKUP, REMAX_ROOT } from '../constants';
import pure from '../utils/pure';
import { reset } from '../instanceId';

class Context {
  data: any;

  requestUpdate(this: any) {
    this.data = {
      [REMAX_ROOT]: pure(this[REMAX_ROOT_BACKUP]),
    };
  }
}

describe('remax render', () => {
  afterEach(() => {
    reset();
  });

  it('render correctly', () => {
    const Page = () => <View className="foo">hello</View>;
    const context = new Context();
    render(<Page />, context);
    expect(context.data).toMatchSnapshot();
  });

  it('insert new element', () => {
    class Page extends React.Component {
      state = {
        list: [1, 3],
      };

      update() {
        this.setState({
          list: [1, 2, 3],
        });
      }

      render() {
        const { list } = this.state;
        return (
          <View>
            {list.map(i => (
              <View key={i}>{i}</View>
            ))}
          </View>
        );
      }
    }

    const context = new Context();
    const page = React.createRef<any>();
    render(<Page ref={page} />, context);
    expect(context.data).toMatchSnapshot();
    page.current.update();
    expect(context.data).toMatchSnapshot();
  });

  it('umount component', () => {
    class Page extends React.Component {
      state = {
        show: true,
      };

      hide() {
        this.setState({ show: false });
      }

      render() {
        return <View>{this.state.show && <View>foo</View>}</View>;
      }
    }
    const context = new Context();
    const page = React.createRef<any>();
    render(<Page ref={page} />, context);
    expect(context.data).toMatchSnapshot();
    page.current.hide();
    expect(context.data).toMatchSnapshot();
  });

  it('renders style', () => {
    const Page = () => <View style={{ width: '100px', height: '100px' }}>hello</View>;
    const context = new Context();
    render(<Page />, context);
    expect(context.data).toMatchSnapshot();
  });
});
