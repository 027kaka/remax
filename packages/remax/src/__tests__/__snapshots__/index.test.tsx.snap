// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remax render insert new element 1`] = `
Object {
  "$$REMAX_ROOT": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "1",
                  "type": Symbol(text),
                },
              ],
              "id": 0,
              "props": Object {},
              "type": "view",
            },
            Object {
              "children": Array [
                Object {
                  "text": "3",
                  "type": Symbol(text),
                },
              ],
              "id": 1,
              "props": Object {},
              "type": "view",
            },
          ],
          "id": 2,
          "props": Object {},
          "type": "view",
        },
      ],
      "type": "root",
    },
  ],
}
`;

exports[`remax render insert new element 2`] = `
Object {
  "$$REMAX_ROOT": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "1",
                  "type": Symbol(text),
                },
              ],
              "id": 0,
              "props": Object {},
              "type": "view",
            },
            Object {
              "children": Array [
                Object {
                  "text": "2",
                  "type": Symbol(text),
                },
              ],
              "id": 0,
              "props": Object {},
              "type": "view",
            },
            Object {
              "children": Array [
                Object {
                  "text": "3",
                  "type": Symbol(text),
                },
              ],
              "id": 1,
              "props": Object {},
              "type": "view",
            },
          ],
          "id": 2,
          "props": Object {},
          "type": "view",
        },
      ],
      "type": "root",
    },
  ],
}
`;

exports[`remax render render correctly 1`] = `
Object {
  "$$REMAX_ROOT": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "text": "hello",
              "type": Symbol(text),
            },
          ],
          "id": 0,
          "props": Object {
            "className": "foo",
          },
          "type": "view",
        },
      ],
      "type": "root",
    },
  ],
}
`;

exports[`remax render umount component 1`] = `
Object {
  "$$REMAX_ROOT": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "text": "foo",
                  "type": Symbol(text),
                },
              ],
              "id": 0,
              "props": Object {},
              "type": "view",
            },
          ],
          "id": 1,
          "props": Object {},
          "type": "view",
        },
      ],
      "type": "root",
    },
  ],
}
`;

exports[`remax render umount component 2`] = `
Object {
  "$$REMAX_ROOT": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": 1,
          "props": Object {},
          "type": "view",
        },
      ],
      "type": "root",
    },
  ],
}
`;
